File: hello-mcp-swift/mcp-client/Sources/Client/OllamaClient.swift
import Foundation

#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// 消息类
public struct OllamaMessage: Codable {
    public let role: String
    public let content: String
    
    public init(role: String, content: String) {
        self.role = role
        self.content = content
    }
}

// 为了向后兼容，保留 Message 别名
public typealias Message = OllamaMessage

/// 工具调用类
public struct ToolCall: Codable {
    public let name: String
    public let arguments: [String: AnyCodable]
    
    public init(name: String, arguments: [String: AnyCodable]) {
        self.name = name
        self.arguments = arguments
    }
}

/// 聊天响应类
public struct ChatResponse {
    public let role: String
    public let content: String
    public let toolCalls: [ToolCall]
    
    public var hasToolCalls: Bool {
        return !toolCalls.isEmpty
    }
}

/// Ollama 客户端
/// 用于与 Ollama API 交互，支持工具调用
public class OllamaClient {
    private let baseURL: String
    private let model: String
    
    public init(baseURL: String = "http://localhost:11434", model: String = "qwen2.5:latest") {
        self.baseURL = baseURL
        self.model = model
    }
    
    /// 发送聊天请求
    public func chat(messages: [Message], tools: [[String: AnyCodable]]) async throws -> ChatResponse {
        print("发送聊天请求到 Ollama: model=\(model), messages=\(messages.count)")
        
        let url = URL(string: "\(baseURL)/api/chat")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 300
        
        let requestBody: [String: Any] = [
            "model": model,
            "stream": false,
            "messages": messages.map { ["role": $0.role, "content": $0.content] },
            "tools": tools.map { tool in
                tool.mapValues { value -> Any in
                    switch value {
                    case .string(let s): return s
                    case .int(let i): return i
                    case .double(let d): return d
                    case .bool(let b): return b
                    case .dictionary(let dict): return dict
                    case .array(let arr): return arr
                    case .null: return NSNull()
                    }
                }
            }
        ]
        
        request.httpBody = try JSONSerialization.data(withJSONObject: requestBody)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw NSError(domain: "OllamaClient", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid response"])
        }
        
        guard httpResponse.statusCode == 200 else {
            throw NSError(domain: "OllamaClient", code: httpResponse.statusCode, 
                         userInfo: [NSLocalizedDescriptionKey: "Ollama API 请求失败: \(httpResponse.statusCode)"])
        }
        
        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        let messageNode = json?["message"] as? [String: Any]
        
        let role = messageNode?["role"] as? String ?? "assistant"
        let content = messageNode?["content"] as? String ?? ""
        
        var toolCalls: [ToolCall] = []
        if let toolCallsArray = messageNode?["tool_calls"] as? [[String: Any]] {
            for toolCallNode in toolCallsArray {
                if let function = toolCallNode["function"] as? [String: Any],
                   let name = function["name"] as? String,
                   let arguments = function["arguments"] as? [String: Any] {
                    let args = arguments.mapValues { AnyCodable($0) }
                    toolCalls.append(ToolCall(name: name, arguments: args))
                }
            }
        }
        
        return ChatResponse(role: role, content: content, toolCalls: toolCalls)
    }
    
    /// 执行工具调用
    public func executeToolCall(toolCall: ToolCall, helloClient: HelloClient) async throws -> String {
        print("执行工具调用: \(toolCall.name), 参数: \(toolCall.arguments)")
        
        var result: String
        
        switch toolCall.name {
        case "get_element":
            if case .string(let name) = toolCall.arguments["name"] {
                result = try await helloClient.getElement(name: name)
            } else {
                result = "{\"error\": \"参数错误\"}"
            }
            
        case "get_element_by_position":
            if case .int(let position) = toolCall.arguments["position"] {
                result = try await helloClient.getElementByPosition(position: position)
            } else if case .double(let position) = toolCall.arguments["position"] {
                result = try await helloClient.getElementByPosition(position: Int(position))
            } else {
                result = "{\"error\": \"参数错误\"}"
            }
            
        default:
            result = "{\"error\": \"未知工具: \(toolCall.name)\"}"
        }
        
        print("工具调用结果: \(result)")
        return result
    }
}

/// AnyCodable 类型用于处理动态 JSON
public enum AnyCodable: Codable {
    case string(String)
    case int(Int)
    case double(Double)
    case bool(Bool)
    case dictionary([String: AnyCodable])
    case array([AnyCodable])
    case null
    
    public init(_ value: Any) {
        if let string = value as? String {
            self = .string(string)
        } else if let int = value as? Int {
            self = .int(int)
        } else if let double = value as? Double {
            self = .double(double)
        } else if let bool = value as? Bool {
            self = .bool(bool)
        } else if let dict = value as? [String: Any] {
            self = .dictionary(dict.mapValues { AnyCodable($0) })
        } else if let array = value as? [Any] {
            self = .array(array.map { AnyCodable($0) })
        } else {
            self = .null
        }
    }
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        
        if let string = try? container.decode(String.self) {
            self = .string(string)
        } else if let int = try? container.decode(Int.self) {
            self = .int(int)
        } else if let double = try? container.decode(Double.self) {
            self = .double(double)
        } else if let bool = try? container.decode(Bool.self) {
            self = .bool(bool)
        } else if let dict = try? container.decode([String: AnyCodable].self) {
            self = .dictionary(dict)
        } else if let array = try? container.decode([AnyCodable].self) {
            self = .array(array)
        } else {
            self = .null
        }
    }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        
        switch self {
        case .string(let value):
            try container.encode(value)
        case .int(let value):
            try container.encode(value)
        case .double(let value):
            try container.encode(value)
        case .bool(let value):
            try container.encode(value)
        case .dictionary(let value):
            try container.encode(value)
        case .array(let value):
            try container.encode(value)
        case .null:
            try container.encodeNil()
        }
    }
}

File: hello-mcp-rust/mcp-client/ollama_client.rs
use anyhow::Result;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::HashMap;
use tracing::{debug, error, info};

use crate::client::HelloClient;

/// 消息类
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Message {
    pub role: String,
    pub content: String,
}

impl Message {
    pub fn new(role: &str, content: &str) -> Self {
        Self {
            role: role.to_string(),
            content: content.to_string(),
        }
    }
}

/// 工具调用类
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolCall {
    pub name: String,
    pub arguments: HashMap<String, Value>,
}

/// 聊天响应类
#[derive(Debug, Clone)]
pub struct ChatResponse {
    pub role: String,
    pub content: String,
    pub tool_calls: Vec<ToolCall>,
}

impl ChatResponse {
    pub fn has_tool_calls(&self) -> bool {
        !self.tool_calls.is_empty()
    }
}

/// Ollama 客户端
/// 
/// 用于与 Ollama API 交互，支持工具调用
pub struct OllamaClient {
    base_url: String,
    model: String,
    http_client: Client,
}

impl OllamaClient {
    pub fn new(base_url: &str, model: &str) -> Self {
        Self {
            base_url: base_url.to_string(),
            model: model.to_string(),
            http_client: Client::builder()
                .timeout(std::time::Duration::from_secs(300))
                .build()
                .unwrap(),
        }
    }

    /// 发送聊天请求
    /// 
    /// # Arguments
    /// * `messages` - 消息列表
    /// * `tools` - 可用工具列表
    /// 
    /// # Returns
    /// 响应消息
    pub async fn chat(&self, messages: &[Message], tools: &[Value]) -> Result<ChatResponse> {
        info!(
            "发送聊天请求到 Ollama: model={}, messages={}",
            self.model,
            messages.len()
        );

        // 构建请求体
        let request_body = json!({
            "model": self.model,
            "stream": false,
            "messages": messages,
            "tools": tools,
        });

        debug!(
            "请求 JSON: {}",
            serde_json::to_string_pretty(&request_body)?
        );

        // 发送请求
        let response = self
            .http_client
            .post(format!("{}/api/chat", self.base_url))
            .json(&request_body)
            .header("Content-Type", "application/json")
            .send()
            .await?;

        let status = response.status();
        debug!("响应状态: {}", status);

        if !status.is_success() {
            let error_text = response.text().await?;
            error!("Ollama API 请求失败: {}", error_text);
            return Err(anyhow::anyhow!("Ollama API 请求失败: {}", status));
        }

        let response_json: Value = response.json().await?;
        debug!("响应内容: {}", serde_json::to_string_pretty(&response_json)?);

        let message_node = &response_json["message"];
        let role = message_node["role"].as_str().unwrap_or("assistant").to_string();
        let content = message_node["content"].as_str().unwrap_or("").to_string();

        // 解析工具调用
        let mut tool_calls = Vec::new();
        if let Some(tool_calls_array) = message_node["tool_calls"].as_array() {
            for tool_call_node in tool_calls_array {
                if let Some(function_node) = tool_call_node["function"].as_object() {
                    let name = function_node["name"].as_str().unwrap_or("").to_string();
                    let arguments = function_node["arguments"]
                        .as_object()
                        .map(|obj| {
                            obj.iter()
                                .map(|(k, v)| (k.clone(), v.clone()))
                                .collect::<HashMap<String, Value>>()
                        })
                        .unwrap_or_default();

                    tool_calls.push(ToolCall { name, arguments });
                }
            }
        }

        Ok(ChatResponse {
            role,
            content,
            tool_calls,
        })
    }

    /// 执行工具调用
    /// 
    /// # Arguments
    /// * `tool_call` - 工具调用信息
    /// * `hello_client` - MCP客户端实例
    /// 
    /// # Returns
    /// 工具执行结果
    pub async fn execute_tool_call(
        &self,
        tool_call: &ToolCall,
        hello_client: &HelloClient,
    ) -> String {
        info!(
            "执行工具调用: {}, 参数: {:?}",
            tool_call.name, tool_call.arguments
        );

        let result = match tool_call.name.as_str() {
            "get_element" => {
                if let Some(name) = tool_call.arguments.get("name").and_then(|v| v.as_str()) {
                    hello_client.get_element(name).await.unwrap_or_else(|e| {
                        json!({"error": e.to_string()}).to_string()
                    })
                } else {
                    json!({"error": "缺少参数: name"}).to_string()
                }
            }
            "get_element_by_position" => {
                if let Some(position) = tool_call.arguments.get("position") {
                    let pos = if position.is_f64() {
                        position.as_f64().unwrap() as i32
                    } else {
                        position.as_i64().unwrap_or(0) as i32
                    };
                    hello_client
                        .get_element_by_position(pos)
                        .await
                        .unwrap_or_else(|e| json!({"error": e.to_string()}).to_string())
                } else {
                    json!({"error": "缺少参数: position"}).to_string()
                }
            }
            _ => json!({"error": format!("未知工具: {}", tool_call.name)}).to_string(),
        };

        info!("工具调用结果: {}", result);
        result
    }
}
